src/features/device-state/routes/state.ts

```diff
@@ -174,121 +347,133 @@ export const state: RequestHandler = async (req, res) => {
-					(image.content_hash != null ? `@${image.content_hash}` : '');
-
-				services[svc.id] = {
-					imageId: image.id,
-					serviceName: svc.service_name,
-					image: formatImageLocation(imgRegistry),
-					// This needs spoken about...
-					running: true,
-					environment,
-					labels,
-				};
-				// Don't send a null contract as this is a waste
-				// of bandwidth (a null contract is the same as
-				// the lack of a contract field)
-				if (image.contract != null) {
-					services[svc.id].contract = image.contract;
-				}
+		// get the release of the main app that this device should run...
+		const release = getReleaseForDevice(device);
 
-				if (
-					composition != null &&
-					composition.services != null &&
-					composition.services[svc.service_name] != null
-				) {
-					const compositionService = composition.services[svc.service_name];
-					// We remove the `build` properly explicitly as it's expected to be present
-					// for the builder, but makes no sense for the supervisor to support
-					delete compositionService.build;
-					services[svc.id] = {
-						...compositionService,
-						...services[svc.id],
-					};
-				}
-			});
-		}
+		// grab the main app for this device...
+		const mainApp =
+			release == null
+				? {
+						name: parentApp.app_name,
+						services: {},
+						networks: {},
+						volumes: {},
+				  }
+				: await buildAppFromRelease(device, parentApp, release, config);
+
+		/**
+		 * Grab the supervisor version for this device, either:
+		 * - the version which should be managing the device
+		 * - the version last reported to be managing the device
+		 * - empty string
+		 */
+		const supervisorVersion: string = (() => {
+			const [supervisorRelease] = device.should_be_managed_by__release;
+
+			if (supervisorRelease?.release_version != null) {
+				return supervisorRelease.release_version;
+			}
 
-		const volumes = composition?.volumes || {};
-		const networks = composition?.networks || {};
+			return device.supervisor_version ?? '';
+		})();
+
+		// grab the system apps for this device...
+		const extraContainers =
+			conf.EXTRA_CONTAINERS.length === 0
+				? []
+				: await sbvrUtils.db.readTransaction!(async (tx) => {
+						const resinApiTx = api.resin.clone({ passthrough: { req, tx } });
+						return await resinApiTx.get({
+							resource: 'application',
+							options: {
+								$select: ['id', 'app_name', 'uuid', 'install_type'],
+								$expand: {
+									application_config_variable: {
+										$select: ['name', 'value'],
+										$orderby: {
+											name: 'asc',
+										},
+									},
+									application_environment_variable: {
+										$select: ['name', 'value'],
+										$orderby: {
+											name: 'asc',
+										},
+									},
+									should_be_running__release: releaseExpand,
+								},
+								$filter: {
+									uuid: {
+										$in: conf.EXTRA_CONTAINERS,
+									},
+									should_be_running__release: {
```

Comment:
this is likely insufficient, as it'd limit supervisors in the state endpoint to only those that are the current latest release in their respective apps.

Reply:
immediately we can work around this by changing this line as follows i believe:
```suggestion
									owns__release: {
```

Reply:
@Page- specifically here i don't think there's a clean way in the same query to fetch both:

1. the release linked to a device in the case of the supervisor (i.e., whatever we find at `should_be_managed_by__release`), AND
1. the latest release in the case of a hostapp extension/hostapp

am i missing something, or does this in fact need to be split into two queries and merged back into a single object?